pkg/report/template.go

```diff
@@ -0,0 +1,141 @@
+)
+
+const (
+	staticDir    = "statics"
+	templatesDir = "templates"
+)
+
+var (
+	//go:embed statics/*
+	staticsFS embed.FS
+	//go:embed templates/*.go.html
+	templatesFS embed.FS
+)
+
+// Generate create a DIBReport on the filesystem.
+func Generate(dibReport DIBReport, dag dag.DAG) error {
+	logrus.Infof("generating html report in the %s folder...", dibReport.GetRootDir())
+	if err := graphviz.GenerateGraph(&dag, dibReport.GetRootDir()); err != nil {
+		return fmt.Errorf("unable to generate graph: %w", err)
+	}
+
+	if err := copyStaticFiles(dibReport); err != nil {
+		return fmt.Errorf("unable to create report static file: %w", err)
+	}
+
+	if err := renderTemplates(dibReport); err != nil {
+		return fmt.Errorf("unable to render report templates: %w", err)
+	}
+
+	return nil
+}
+
+// copyStaticFiles iterate recursively on the "statics" embed filesystem and copy it inside the report folder.
+func copyStaticFiles(dibReport DIBReport) error {
+	return fs.WalkDir(staticsFS, staticDir, func(itemPath string, itemEntry fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if itemEntry.IsDir() {
+			if err = os.MkdirAll(path.Join(dibReport.GetRootDir(), itemPath), 0o755); err != nil {
+				return err
+			}
+			return nil
+		}
+
+		data, err := fs.ReadFile(staticsFS, itemPath)
+		if err != nil {
+			return err
+		}
+
+		err = os.WriteFile(path.Join(dibReport.GetRootDir(), itemPath), data, 0o600)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+// renderTemplates compile & render DIBReport templates files then create it inside the report folder.
+func renderTemplates(dibReport DIBReport) error {
+	data := make(map[string]interface{})
+	data["buildUID"] = dibReport.Name
+	data["generationDate"] = dibReport.GenerationDate
+	data["buildReport"] = dibReport.BuildReports
+
+	// Generate index.html
+	if err := dibReport.executeAndWrite("index", data); err != nil {
+		return err
+	}
+
+	// Generate graph.html
+	if err := dibReport.executeAndWrite("graph", nil); err != nil {
+		return err
+	}
+
+	// Generate build.html
+	buildLogsData, err := parseBuildLogs(dibReport.GetBuildLogsDir())
+	if err != nil {
+		return err
+	}
+
+	if err := dibReport.executeAndWrite("build", buildLogsData); err != nil {
+		return err
+	}
+
+	// Generate scan.html
+	if err := dibReport.executeAndWrite("scan", nil); err != nil {
+		return err
+	}
+
+	// Generate test.html
+	if err := dibReport.executeAndWrite("test", nil); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// parseBuildLogs iterate over the build logs dir and read each files.
+// Then, it put in a map that will be used in Go template.
+func parseBuildLogs(path string) (map[string]string, error) {
+	buildLogsData := make(map[string]string)
+	err := filepath.Walk(path, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		imageName := GetDockerImageNameFromBuildLogFile(info.Name())
```

Comment:
Why do we need to parse the logs to get the image name ? The image name should be available in each `BuildReport`